### PR TITLE
Support semver

### LINKED ISSUE #1439: Error message for versions like 1.4.0-pre.0
### Describe the problem and steps to reproduce it:
Error message for a bad version format is too cryptic.

### What happened?
Used a version like 1.4.0-pre.0 for a beta version of my addon.
Got the error: "/version" should match format "versionString"

### What did you expect to happen?
Either no error, or an error explaining the format

### PR DESCRIPTION
**Pull request by asamuzaK**
Fixes  #1439, #1172

### CODE REVIEW

**Comment by asamuzaK** [comment]
Notes:
* `'0.1.12dev-cb31c51'`,
* `'4.1.1dev-abcdef1'`,
moved to invalid string, because `dev` is not allowed for beta string.

* `'1.0.0-beta2'`
moved to valid string, because it is valid semver.

**Comment by asamuzaK** [comment]
https://github.com/mozilla/addons-linter/issues/1439#issuecomment-324538056

I'll rewrite PR when how to treat '0.1.12dev-cb31c51', '4.1.1dev-abcdef1' should be is decided.

Another question from me.
Currently, isToolkitVersionString() === isValidVersionString() (Some kind of an alias).
But, when we support semver, valid version string is not always equal to toolkit version string.

Should we keep isToolkitVersionString an alias? (current implement)

Or should we treat those function to match there name, like below? (may break some functions)

```
const TOOLKIT_VERSION_REGEX = ...;
function isValidVersionString(version) {
  if (TOOLKIT_VERSION_REGEX.test(version)) {
    return true;
  }
  const semverReg = ...;
  return semverReg.test(version);
}
function isToolkitVersionString(version) {
  return TOOLKIT_VERSION_REGEX.test(version);
}
```

**Comment by EnTeQuAk** [comment]
`isToolkitVersionString` and `isValidVersionString` interestingly call each other which I think feels wrong. Imho `isValidVersionString` should call `isToolkitVersionString` for an early return and then go on parse the rest. The usage of `isToolkitVersionString` in `src/parsers/manifestjson.js` should be replaced by `isValidVersionString`.

File: src/schema/formats.js [hunk 1]
```diff
@@ -1,48 +1,27 @@
...
+  // permissive than Chrome) to allow Beta addons, per:
+  // https://developer.mozilla.org/en-US/Add-ons/AMO/Policy/Maintenance
+  // We also support Semantic Versioning:
+  // http://semver.org/
+  const chromePart = '(?:0|[1-9]\\d{0,3}|[1-5]\\d{4}|6(?:[0-4]\\d{3}|5(?:[0-4]\\d{2}|5(?:[0-2]\\d|3[0-5]))))';
+  const chromeVer = `${chromePart}(?:\\.${chromePart}){0,3}`;
+  const betaPart = '(?:a(?:lpha)?|b(?:eta)?|pre|rc)\\d*';
+  const addonVer = `${chromeVer}(?:${betaPart})?`;
+  const semVer = '(?:0|[1-9]\\d*)(?:\\.(?:0|[1-9]\\d*)){2}(?:-[A-Za-z\\d\\-]+(?:\\.[A-Za-z\\d\\-]+)*)?(?:\\+[A-Za-z\\d\\-]+(?:\\.[A-Za-z\\d\\-]+)*)?';
+  return (new RegExp(`^(?:${addonVer}|${semVer})$`)).test(version);
```
Comment (wagnerand) [inline src/schema/formats.js]: I think those regular expressions should use what Firefox accepts, not more and not less.

Comment (asamuzaK) [reply src/schema/formats.js]: I am not a native English speaker, so I can not understand a bit of the meaning of what you are pointing out.
What does "use what Firefox accepts" mean?

Comment (wagnerand) [reply src/schema/formats.js]: We should use the format that Firefox supports.

Comment (asamuzaK) [reply src/schema/formats.js]: You mean only Toolkit?

Comment (asamuzaK) [reply src/schema/formats.js]: Chrome version format always matches Toolkit. (&lt;number-a&gt;)
SemVer should match Toolkit too.
For example, `1.0.0-a1b1`.
According to [Toolkit version format | MDN](https://developer.mozilla.org/en-US/docs/Toolkit_version_format "Toolkit version format | MDN"), this version looks like it doesn't match Toolkit version format.
In 3rd version part (`0-a1b1`), &lt;number-a&gt; is `0` &lt;string-b&gt; is `-a` &lt;number-c&gt; is `1` &lt;string-d&gt; is `b`, last `1` does not match.
But according to [nsIVersionComparator - Mozilla | MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIVersionComparator "nsIVersionComparator - Mozilla | MDN"), it says
&gt; &lt;number-a&gt;&lt;string-b&gt;&lt;number-c&gt;&lt;string-d (everything else)&gt;
If so, then &lt;string-d&gt; becomes `b1`, and it matches Toolkit?

**Comment by asamuzaK** [comment]
Closing this PR. See https://github.com/mozilla/addons-linter/issues/1439#issuecomment-326158984

**Pull request closed without merging**